onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -57,12 +60,29 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
   std::unordered_set<const NodeArg*> cpu_args;
   std::unordered_set<NodeIndex> provider_nodes;
   std::unordered_map<NodeIndex, const KernelCreateInfo*> node_to_kernel;
+  std::unordered_set<NodeIndex> cpu_kernel_available;
 
   for (auto& node_id : tentative_nodes) {
     provider_nodes.insert(node_id);
     const Node* node = graph.GetNode(node_id);
 
     const KernelCreateInfo* kernel_info = nullptr;
+
+    // Get the CPU kernel availability for this node
+    KernelRegistryManager mgr;
+    ExecutionProviders cpu_ep;
+    CPUExecutionProviderInfo epi{false};
+    ORT_ENFORCE(cpu_ep.Add(kCpuExecutionProvider, onnxruntime::make_unique<CPUExecutionProvider>(epi)).IsOK());
+    ORT_ENFORCE(mgr.RegisterKernels(cpu_ep).IsOK());
+    std::vector<const KernelRegistry*> cpu_kernel_registries = mgr.GetKernelRegistriesByProviderType(kCpuExecutionProvider);
```

Comment:
This kernel registry should be built once outside the loop. 
ORT has an implicit rule that all the ops would have an CPU kernel, so I am wondering if this check is necessary.

Reply:
Did you encounter a case where a node is placed on CPU, but the cpu kernel is missing?

Reply:
We don't have CPU kernels for some contrib ops like IsAllFinite. Some tests in ORTTrainer were failing as the output of IsAllFinite was consumed by AdamOptiomizer/LambOptimizer, which bind the "do_update" input to CPU which is produced by "IsAllFinite".